src/KeepAlive.cpp

```diff
@@ -152,9 +152,18 @@ void etcd::KeepAlive::refresh()
       std::cerr << "keepalive timer error: " << error << ", " << error.message() << std::endl;
 #endif
     } else {
-      this->stubs->call->Refresh();
-      // trigger the next round;
-      this->refresh();
+      if (this->continue_next) {
+        auto resp = this->stubs->call->Refresh();
+        if (!resp.is_ok()) {
+          throw std::runtime_error("Failed to refresh lease: error code: " + std::to_string(resp.error_code()) +
+                                   ", message: " + resp.error_message());
+        }
+        if (resp.value().ttl() == -1) {
+          throw std::runtime_error("Failed to refresh lease due to expiration: the new TTL is -1.");
```

Comment:
This  throwing of the same exception for both errors is not good. We cannot distinguish in practice between a failure of the connection - after which we should create a new KeepAlive - and expiry of the lease after which we need a new lease.

Reply:
>  It's important to be able to distinguish between the lease expiry and the connection failure.

Agree about distinguishing those two kinds of failure, but I would like to throw different kinds of exception (`std::runtime_error` and `std::invalid_argument` or `std::out_of_range`). I don't want to makes the `KeepAlive` so complicate for users. After adding a another handler, there would be too many ctor combinations and dispatch logic.